authorization/src/main/kotlin/org/modelix/authorization/PermissionManagementPage.kt

```diff
@@ -0,0 +1,254 @@
+import kotlinx.html.body
+import kotlinx.html.br
+import kotlinx.html.div
+import kotlinx.html.h1
+import kotlinx.html.head
+import kotlinx.html.hiddenInput
+import kotlinx.html.postForm
+import kotlinx.html.style
+import kotlinx.html.submitInput
+import kotlinx.html.table
+import kotlinx.html.td
+import kotlinx.html.textInput
+import kotlinx.html.th
+import kotlinx.html.tr
+import org.modelix.authorization.permissions.PermissionParts
+import org.modelix.authorization.permissions.PermissionSchemaBase
+
+fun Route.installPermissionManagementHandlers() {
+    route("permissions") {
+        get("manage") {
+            call.respondHtml {
+                buildPermissionManagementPage(call, application.plugin(ModelixAuthorization))
+            }
+        }
+        post("grant") {
+            val formParameters = call.receiveParameters()
+            val userId = formParameters["userId"]
+            val roleId = formParameters["roleId"]
+            require(userId != null || roleId != null) { "userId or roleId required" }
+            val permissionId = requireNotNull(formParameters["permissionId"]) { "permissionId not specified" }
+
+            // a user can grant his own permission to other users
+            checkPermission(PermissionParts.fromString(permissionId))
+
+            if (userId != null) {
+                application.plugin(ModelixAuthorization).config.accessControlPersistence.update {
+                    it.withGrantToUser(userId, permissionId)
+                }
+            }
+            if (roleId != null) {
+                application.plugin(ModelixAuthorization).config.accessControlPersistence.update {
+                    it.withGrantToRole(roleId, permissionId)
+                }
+            }
+            call.respond("Granted $permissionId to ${userId ?: roleId}")
+        }
+        post("remove-grant") {
+            call.checkPermission(PermissionSchemaBase.permissionData.write)
+            val formParameters = call.receiveParameters()
+            val userId = formParameters["userId"]
+            val roleId = formParameters["roleId"]
+            require(userId != null || roleId != null) { "userId or roleId required" }
+            val permissionId = requireNotNull(formParameters["permissionId"]) { "permissionId not specified" }
+            if (userId != null) {
+                application.plugin(ModelixAuthorization).config.accessControlPersistence.update {
+                    it.withoutGrantToUser(userId, permissionId)
+                }
+            }
+            if (roleId != null) {
+                application.plugin(ModelixAuthorization).config.accessControlPersistence.update {
+                    it.withoutGrantToUser(roleId, permissionId)
+                }
+            }
+            call.respond("Removed $permissionId to ${userId ?: roleId}")
+        }
+    }
+}
+
+fun HTML.buildPermissionManagementPage(call: ApplicationCall, pluginInstance: ModelixAuthorizationPluginInstance) {
```

Comment:
thought: The rules for what is shown in the tables seem odd. (But not problematic.)

* Tables with granted permissions, shows only entries with permissions the viewing users has.
  * This is odd in the following scenario: User A "owns" a repository. User A gives user B the rights to read the repository. Now user B can inspect, who can read the repository but cannot see, who can write. Also user B gets a button to remove the grant.
  


* The table with denied permission, shows only entries with permissions the viewing users has.
  * So a user can only see denied permission, to resources, they have access to.
  * This is odd in the following scenario: User A "owns" a repository. User A gives user B the rights to read the repository. Now user B can audit denied read attempts on the repository.

Reply:
At least it's consistent to who is allowed to grant permissions. Removing grants is only allowed by cluster admins. The button should be hidden/disabled for other users.

Reply:
Yes, it is consistent. And therefore fine.